src/mpid/common/genq/mpidu_genq_shmem_queue.c

```diff
@@ -4,28 +4,48 @@
  */
 
 #include "mpidimpl.h"
-#include "mpidu_genq_shmem_pool.h"
-#include "mpidu_genq_shmem_queue.h"
-#include "mpidu_genqi_shmem_types.h"
 
-int MPIDU_genq_shmem_queue_init(MPIDU_genq_shmem_queue_t queue, MPIDU_genq_shmem_pool_t pool,
-                                int flags)
+int MPIDU_genq_shmem_queues_alloc(int num_of_queues, MPIDU_genq_shmem_queues_t * queues)
 {
-    int rc = MPI_SUCCESS;
+    int mpi_errno = MPI_SUCCESS;
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDU_GENQ_SHMEM_QUEUE_ALLOC);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDU_GENQ_SHMEM_QUEUE_ALLOC);
+
+    size_t total_size = (size_t) num_of_queues * sizeof(genqi_shmem_queue_u);
+
+    mpi_errno = MPIDU_Init_shm_alloc(total_size, (void *) &queues->ptr);
+    MPIR_ERR_CHECK(mpi_errno);
+
+  fn_exit:
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPIDU_GENQ_SHMEM_QUEUE_ALLOC);
+    return mpi_errno;
+  fn_fail:
+    goto fn_exit;
+}
 
+int MPIDU_genq_shmem_queues_free(MPIDU_genq_shmem_queues_t queues)
+{
+    return MPIDU_Init_shm_free(queues.ptr);
+}
+
+/* initialize queue on a existing shared memeory */
+int MPIDU_genq_shmem_queue_init(MPIDU_genq_shmem_queues_t queues, int queue_idx, int flags)
```

Comment:
If we remove the "queues_alloc", we can revert this to simply init queue on a given address.

Reply:
I am removing this instead -- (making it internal at least). I am not convinced that we ever need break the abstraction and let user init a bare queue on a piece of shared memory.